src/commands/ai.cr

```diff
@@ -0,0 +1,116 @@
+require "anthropic"
+require "xml"
+
+module Wax::Commands::AI
+  extend self
+
+  Claude = Anthropic::Client.new
+
+  def call(args : Array(String))
+    if args.empty?
+      STDERR.puts "Must supply a prompt"
+      exit 1
+    end
+    puts "Analyzing the repo to determine how to accomplish this."
+    prompt = args.first
+
+    # TODO: Figure out how to reduce the necessary context
+    files = Dir["README.md", "src/**/*.cr", "views/**/*.ecr", "spec/**/*_spec.cr", "db/migrations/**/*.sql"].map do |filename|
+      FileData.new(
+        path: filename.lchop(ENV["PWD"]),
+        contents: File.read(filename),
+      )
+    end
+
+    file_contents = String.build do |str|
+      str.puts "We currently have these files:"
+      str.puts
+
+      XML.build_fragment str do |xml|
+        files.each do |file|
+          xml.element "file", path: file.path do
+            xml.text file.contents
+          end
+        end
+      end
+    end
+
+    puts Claude.messages.create(
+      model: Anthropic.model_name(:sonnet),
+      # model: Anthropic.model_name(:haiku),
+      messages: [
+        Anthropic::Message.new(content: Array(Anthropic::MessageContent){
+          Anthropic::Text.new(file_contents),
+          Anthropic::Text.new(prompt),
+        }),
+      ],
+      tools: [
+        WriteFiles,
+        DeleteFiles,
+      ],
```

Comment:
I wonder if, to improve token consumption, we should also implement an `ApplyDiffs` tool here.

A constraint imposed by Claude 3.5 Sonnet is that the maximum response is 8192 tokens and `WriteFiles` requires the complete file contents in the tool use response even if it only changes a single line (such as adding a link in a template to use some new functionality being added), so that imposes a hard limit on the size of any single file Claude can generate for us with `WriteFiles`. If we have to update one line in multiple files that are each 4k-8k tokens, we can only update one file per API request, requiring us to send the complete context every single time. That's not an efficient use of our token budget.

Diffs can represent the creation, modification, and deletion of files, but creation and deletion might be more efficient with `WriteFiles` and `DeleteFiles`, respectively — a diff for creation contains additional diff metadata and a diff representing deletion of a file has to include a deletion of every line in the file, whereas the `DeleteFiles` tool is just about the most efficient it can feasibly be.

Another point in favor of an `ApplyDiffs` tool is that, since the Anthropic client will keep invoking tools until it executes everything the LLM has planned out, we'd no longer be constrained to an 8k-token file size. Personally, I think having that constraint *can* be a good thing, but I'm not sure what to do if a file *can't* satisfy that constraint. With `ApplyDiffs`, the LLM might be able to write out a chunk of the file with `WriteFiles` and then invoke `ApplyDiffs` to add to it. I'm not *sure* it would do that (because nobody really knows what an LLM will do, which is admittedly kinda scary), but in my own projects this PR is already doing some things I hadn't considered it would, such as splitting the updates across multiple API requests when it can't fit the full set of changes into one response.

Reply:
Just tried this and it kept generating patches that `git apply` couldn't apply. Someone more familiar with git diffs than I am will probably need to look into it.